crates/next-error-code-swc-plugin/src/lib.rs

```diff
@@ -0,0 +1,275 @@
+        })
+        .to_string()
+        .hash(&mut hasher);
+        let hash = format!("{:08x}", hasher.finish());
+
+        // Concatenate the commit hash and the hash
+        let error_code = format!("E{}{}", self.commit_hash, hash);
+
+        // Save to JSON file
+        let error_metadata = serde_json::json!({
+            "file_path": self.file_path,
+            "error_message": error_message,
+            "occurrence_count": self.string_occurrences[&error_message]
+        })
+        .to_string();
+        let mut retries = 3;
+
+        // Do not do FS operations in unit tests
+        if cfg!(test) {
+            return error_code;
+        }
+
+        loop {
+            if self.mode == "check" {
+                let error_code_path = format!("cwd/error_codes/{}.json", hash);
+                if !std::path::Path::new(&error_code_path).exists() {
+                    panic!(
+                        "ERROR: File {} does not exist.\n\nREQUIRED ACTION:\n1. Run `pnpm \
+                         build`\n2. Commit all file changes from \
+                         /packages/next/error-codes\n\nThis is required to maintain error code \
+                         consistency.",
+                        format!("/packages/next/error_codes/{}.json", hash)
+                    );
+                }
+                return error_code;
+            }
+
+            if self.mode != "generate" {
+                panic!("Mode must be 'generate', got '{}'", self.mode);
+            }
+
+            fs::create_dir_all("cwd/error_codes").unwrap_or_else(|e| {
+                panic!("Failed to create error_codes directory: {}", e);
+            });
+
+            match fs::write(format!("cwd/error_codes/{}.json", hash), &error_metadata) {
+                Ok(_) => break,
+                Err(e) => {
+                    retries -= 1;
+                    if retries == 0 {
+                        panic!("Failed to write error metadata after 3 retries: {}", e);
+                    }
+                }
+            }
+        }
+
+        error_code
+    }
+}
+
+impl VisitMut for TransformVisitor {
+    fn visit_mut_expr(&mut self, expr: &mut Expr) {
+        expr.visit_mut_children_with(self);
+
+        let mut code: Option<String> = None;
+        let mut new_error_expr: Option<&NewExpr> = None;
+
+        if let Expr::New(new_expr) = expr {
+            if let Expr::Ident(ident) = &*new_expr.callee {
+                if ident.sym.to_string() == "Error" {
+                    new_error_expr = Some(new_expr);
+
+                    if let Some(args) = &new_expr.args {
+                        if let Some(first_arg) = args.first() {
+                            code = Some(self.get_error_code(&first_arg));
+                        }
+                    }
+                }
+            }
+        }
+
+        if let Some(code) = code {
+            if let Some(new_error_expr) = new_error_expr {
+                *expr = Expr::Call(CallExpr {
```

Comment:
i'd consider adding a comment saying what this expression is supposed to be, like

```suggestion
                // Object.assign($expr, { __NEXT_ERROR_CODE: $code })
                *expr = Expr::Call(CallExpr {
```
reading raw ASTs isn't exactly easy